lib/charms/tls_certificates_interface/v3/tls_certificates.py

```diff
@@ -687,6 +691,68 @@ def _get_certificate_expiry_time(certificate: str) -> Optional[datetime]:
         return None
 
 
+def _get_certificate_validity_start_time(certificate: str) -> Optional[datetime]:
+    """Extract expiry time from a certificate string.
+
+    Args:
+        certificate (str): x509 certificate as a string
+
+    Returns:
+        Optional[datetime]: Time when this certificate is valid from or None
+    """
+    try:
+        certificate_object = x509.load_pem_x509_certificate(data=certificate.encode())
+        return certificate_object.not_valid_before_utc
+    except ValueError:
+        logger.warning("Could not load certificate.")
+        return None
+
+
+def _calculate_expiry_notification_time(
```

Comment:
```suggestion
def calculate_expiry_notification_time(
```

This is more of a public method now, especially from the rquirer class point of view and the fact we aim tests directly at it.

Reply:
Alright, I think this applies to `get_certificate_validity_start_time` and `get_certificate_expiry_time` too.
done.